### PR TITLE
Switch microk8s enable wrapper to Python script

### LINKED ISSUE #1446: Switch to unix-style argument handling for addon enabling
This is a proposal to switch to regular unix-style argument handling when enabling addons. Briefly, instead of this:

    microk8s.enable kubeflow:"bundle=lite"

switch to this:

    microk8s.enable kubeflow --bundle lite

The big issue to solve is that right now, users can enable multiple addons at once like this:

    microk8s.enable gpu kubeflow

So, we run into an issue of how to handle argument parsing with multiple addons at once. The easiest thing to do would be to just disallow enabling multiple addons at once when using unix-style argument handling. The only issue with this is that an addon may want positional arguments. For example, say the `kubeflow` addon allowed a positional argument of which bundle size to deploy, like this:

    microk8s.enable kubeflow lite

If the user types `microk8s.enable kubeflow gpu dns`, do we parse that as the user wanting 3 addons enabled, or only `kubeflow` and `dns` enabled, and they mistakenly typed `gpu` instead of `full`?

However, if we stick to only parameter flags such as `--foo bar`, it should be possible to implement unix-style argument handling in a backwards-compatible way. In pseudocode, we could have logic like this:

```python
if all(a.split(':')[0] in ADDONS for a in ARGS):
    # We have a command line made entirely of addon names, with or without params
    # e.g. `microk8s.enable foo bar:"123" baz`
    for addon in ARGS:
        enable(addon)
else:
    # Treat as unix-style arguments, where enabling multiple addons at once is disallowed
    # Positional arguments are also disallowed
    ...
```

As a further proposal, I think we should move to [click](https://click.palletsprojects.com/en/7.x/), which will make it easy to create a cohesive CLI experience, as it automatically handles stuff such as argument parsing and `--help` flags.

### PR DESCRIPTION
**Pull request by knkski**
For now, it's just a clone of the old bash script, with the ability to handle unix-style flag arguments. Using click however, will allow expanding the enable scripts to be full click subcommands, instead of just bash scripts.

Fixes #1446

### CODE REVIEW

File: microk8s-resources/wrappers/microk8s-enable.wrapper [hunk 1]
```diff
@@ -1,64 +1,60 @@
...
-    exit 2
-  fi
-done
-exit $result
+    if all(a.split(':')[0] in ADDONS for a in addons) and len(addons) > 1:
+        for addon in addons:
+            addon, *args = addon.split(':')
+            run([PARENT / ('actions/enable.%s.sh' % addon)] + args.split(';'))
+    else:
+        addon, *args = addons[0].split(':')
```
Comment (johnsca) [inline microk8s-resources/wrappers/microk8s-enable.wrapper:26]: Seems like this would blow up with an IndexError if the user calls this without any args. Should probably have a check for an empty addons list and report a reasonable error.

Comment (knkski) [reply microk8s-resources/wrappers/microk8s-enable.wrapper:26]: Added `required=True` which forces at least one argument to be passed, otherwise help text is displayed.

**Comment by balchua** [comment]
@knkski do we drop the checking for addons if they are already running?  If yes, i think this is required, sometimes users manually edit the resource, and when the user enabled it again, it loses their previous changes.

**Comment by knkski** [comment]
@balchua: The use case for dropping that check is a result of #1397. If a user enables kubeflow and it either fails or the `kubeflow` namespace is later deleted, a user will be unable to either enable or disable kubeflow.

It would be nice to have `microk8s.enable kubeflow` support redeploying in case of a failed enable attempt, but maybe there's a better way to handle that. @ktsakalozos, @balchua: thoughts? cc @RFMVasconcelos, who I think also had opinions on that.

**Comment by balchua** [comment]
Thanks @knkski for the info. Instead of dropping that altogether, maybe we can consider adding `idempotent: true` or `false` into the `addon-list.yaml`?  If `idempotent` is `true` then we dont reapply otherwise we reapply.
Another option i can think of is an option `--force` that users can pass to the enable action.  Example:
`microk8s enable --force kubeflow`
These are just options on top of my head.

**Comment by ktsakalozos** [comment]
> sometimes users manually edit the resource, and when the user enabled it again, it loses their previous changes.

The enable script used to reapply the manifests without checking if the addon was already enabled. This was causing some confusion [2] and we had to fix it [1].

@knkski kubeflow enables the dns addon but I also think it expects the forward dns to be correctly set. Will the kubeflow deploy in an environment where 8.8.8.8 is not reachable?

[1] https://github.com/ubuntu/microk8s/pull/927
[2] https://github.com/ubuntu/microk8s/issues/958

**Comment by balchua** [comment]
So does it mean that `enable` command do not need to check whether it is enabled or not?

**Comment by ktsakalozos** [comment]
Sorry I was not clear. We need to keep the current behavior of not re-applying the addon manifests. We have to respect any configuration users have on the running workloads.

**Comment by balchua** [comment]
Aaa ok. Thanks for clarifying.

**Comment by knkski** [comment]
I've updated this PR so that it keeps the current behavior of not enabling already-enabled addons, for everything but Kubeflow, which has it's own logic for handling that. Namely, it will allow disabling Kubeflow even if Kubeflow is already disabled, which lets us fix #1397. 

I didn't create a `idempotent: true` section in `addon-list.yaml`, opting instead for a hardcoded check for the kubeflow addon, but I could add that if desired.

**Comment by balchua** [comment]
Thanks @knkski.  Im good with that.  The unix style args is nice.

File: microk8s-resources/wrappers/microk8s-disable.wrapper [hunk 2]
```diff
@@ -1,76 +1,14 @@
...
-    echo
-    actions="$(find "${SNAP}/actions" -name '*.yaml' -or -name 'disable.*.sh')"
-    actions="$(echo "$actions" | sed -e 's/.*[/.]\([^.]*\)\..*/\1/' | sort | uniq)"
-    for action in $actions; do
-        echo "  $action"
-    done
-    exit 0
-fi
-
-exit_if_stopped
```
Comment (ktsakalozos) [inline microk8s-resources/wrappers/microk8s-disable.wrapper:33]: I think we should keep this check. If microk8s is stopped we do not want to try to enable anything, right?

**Comment by ktsakalozos** [comment]
@knkski there are a few conflicts in this PR. Do you think you could address them so we get this PR ready for 1.20? Thank you.

File: scripts/wrappers/common/utils.py [hunk 3]
```diff
@@ -211,3 +225,64 @@ def set_service_expected_to_start(service, start=True):
...
+    else:
+        addon, *args = addons[0].split(':')
+
+        if addon in xabled_addons and addon != 'kubeflow':
+            click.echo("Addon %s is already %sd." % (addon, action))
+            sys.exit(0)
+
+        if addon not in existing_addons:
+            click.secho("Addon `%s` not found." % addon, fg='red', err=True)
+            click.echo("The available addons are:\n - %s" % '\n - '.join(existing_addons), err=True)
```
Comment (ktsakalozos) [inline scripts/wrappers/common/utils.py:260]: I am not so sure about this. If we print the list of addons here I easily see a request to put descriptions next to the addon names.

File: scripts/wrappers/common/utils.py [hunk 4]
```diff
@@ -102,13 +113,10 @@ def get_dqlite_info():
...
-    clusterLockFile = os.path.expandvars("${SNAP_DATA}/var/lock/clustered.lock")
-    if os.path.isfile(clusterLockFile):
-        print(
-            "This MicroK8s deployment is acting as a node in a cluster. "
-            "Please use the microk8s status on the master."
-        )
-        exit(0)
+    if (snap_data() / 'var/lock/clustered.lock').exists():
+        click.echo('This MicroK8s deployment is acting as a node in a cluster.')
+        click.echo('Please use `microk8s enable` on the master.')
```
Comment (ktsakalozos) [inline scripts/wrappers/common/utils.py:118]: Based on the error message, I assume method was used by `microk8s.status`.

File: scripts/wrappers/common/utils.py [hunk 5]
```diff
@@ -211,3 +225,64 @@ def set_service_expected_to_start(service, start=True):
...
+
+        if addon not in existing_addons:
+            click.secho("Addon `%s` not found." % addon, fg='red', err=True)
+            click.echo("The available addons are:\n - %s" % '\n - '.join(existing_addons), err=True)
+            sys.exit(1)
+
+        if args and addons[1:]:
+            click.secho(
+                click.style(
+                    "Can't pass string arguments and flag arguments simultaneously!\n", fg='red'
```
Comment (ktsakalozos) [inline scripts/wrappers/common/utils.py:266]: I like what you are doing here, error messages in red. Could we do this across all microk8s in a separate PR? If we do this coloring only here it will look odd.

**Comment by ktsakalozos** [comment]
Just reverted this PR, there are a few things (probably minor) that will need a bit more attention.

On enable/disable we request the status of each addon [1]. We should first make sure the API server is running or else the get status fails with:
```
14:55:29 2020-11-27 at 12:55:29 | INFO + /snap/microk8s/1829/actions/enable.cilium.sh
14:55:29 2020-11-27 at 12:55:29 | INFO Traceback (most recent call last):
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/scripts/wrappers/enable.py", line 32, in <module>
14:55:29 2020-11-27 at 12:55:29 | INFO enable(prog_name='microk8s enable')
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/usr/lib/python3/dist-packages/click/core.py", line 716, in __call__
14:55:29 2020-11-27 at 12:55:29 | INFO return self.main(*args, **kwargs)
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/usr/lib/python3/dist-packages/click/core.py", line 696, in main
14:55:29 2020-11-27 at 12:55:29 | INFO rv = self.invoke(ctx)
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/usr/lib/python3/dist-packages/click/core.py", line 889, in invoke
14:55:29 2020-11-27 at 12:55:29 | INFO return ctx.invoke(self.callback, **ctx.params)
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/usr/lib/python3/dist-packages/click/core.py", line 534, in invoke
14:55:29 2020-11-27 at 12:55:29 | INFO return callback(*args, **kwargs)
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/scripts/wrappers/enable.py", line 25, in enable
14:55:29 2020-11-27 at 12:55:29 | INFO enabled_addons, _ = get_status(get_available_addons(get_current_arch()), True)
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/scripts/wrappers/status.py", line 157, in get_status
14:55:29 2020-11-27 at 12:55:29 | INFO kube_output = kubectl_get("all")
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/scripts/wrappers/common/utils.py", line 169, in kubectl_get
14:55:29 2020-11-27 at 12:55:29 | INFO return run("kubectl", kubeconfig, "get", cmd, "--all-namespaces", die=False)
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/scripts/wrappers/common/utils.py", line 39, in run
14:55:29 2020-11-27 at 12:55:29 | INFO result.check_returncode()
14:55:29 2020-11-27 at 12:55:29 | INFO File "/snap/microk8s/1829/usr/lib/python3.5/subprocess.py", line 659, in check_returncode
14:55:29 2020-11-27 at 12:55:29 | INFO self.stderr)
14:55:29 2020-11-27 at 12:55:29 | INFO subprocess.CalledProcessError: Command '('kubectl', '--kubeconfig=/var/snap/microk8s/1829/credentials/client.config', 'get', 'all', '--all-namespaces')' returned non-zero exit status 1
```

Similar case is when we enable/disable addons. We used to wait for up to 30 seconds for the API server to get ready [2]. This way if an addon enables/disables the apiserver we would not fail in [3].

The last thing that needs some mending is the `--help`. Most of the addons do not have a `--help` so we are pushing the users to this wrong path:
```
ubuntu@ip-172-31-40-63:~$ microk8s.disable --help
Usage: microk8s disable [OPTIONS] ADDONS...

  Disables one or more MicroK8s addons.

  For a list of available addons, run `microk8s status`.

  To see help for individual addons, run:

      microk8s disable ADDON -- --help

Options:
  --help  Show this message and exit.
ubuntu@ip-172-31-40-63:~$ microk8s.disable dns -- --help
Disabling DNS
Reconfiguring kubelet
Removing DNS manifest
serviceaccount "coredns" deleted
configmap "coredns" deleted
deployment.apps "coredns" deleted
service "kube-dns" deleted
clusterrole.rbac.authorization.k8s.io "coredns" deleted
clusterrolebinding.rbac.authorization.k8s.io "coredns" deleted
DNS is disabled
```
We should get back the old help message.


[1] https://github.com/ubuntu/microk8s/pull/1467/files#diff-44f8a77098734624eb59a3c59e17024d622e15fb9927eb4d6f3583ac724ffe25R25 
[2] https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/wrappers/microk8s-enable.wrapper#L31
[3] https://github.com/ubuntu/microk8s/pull/1467/files#diff-cea8894ac12cb68765874a03fba91e79d22c596d86633886b5c02abe52829448R246
